Add CAF::Log option to record the PID

Look at the following snippet from the `ncd.log` file:

```
2015/07/06-14:06:26 [INFO] running component: nss
2015/07/06-14:06:26 ---------------------------------------------------------
2015/07/06-14:06:35
2015/07/06-14:06:35 ------------------------------------------------------------
2015/07/06-14:06:35 [INFO] NCM-NCD version 1.4.1 started by root at: Mon Jul  6 14:06:35 2015
2015/07/06-14:06:39 [WARN] signal handler: received signal: INT
2015/07/06-14:06:39 [ERROR] ncd exiting gracefully after signal hit.
2015/07/06-14:09:52
2015/07/06-14:09:52 =========================================================
2015/07/06-14:09:52
2015/07/06-14:09:52 [OK]   0 errors, 0 warnings executing configure
```

With a little bit of guesswork and intuition, you can figure out what’s going on.  But it would be easier to read if the PID of the `ncm-ncd` process was included, like this:

```
[1273] 2015/07/06-14:06:26 [INFO] running component: nss
[1273] 2015/07/06-14:06:26 ---------------------------------------------------------
[1280] 2015/07/06-14:06:35
[1280] 2015/07/06-14:06:35 ------------------------------------------------------------
[1280] 2015/07/06-14:06:35 [INFO] NCM-NCD version 1.4.1 started by root at: Mon Jul  6 14:06:35 2015
[1280] 2015/07/06-14:06:39 [WARN] signal handler: received signal: INT
[1280] 2015/07/06-14:06:39 [ERROR] ncd exiting gracefully after signal hit.
[1273] 2015/07/06-14:09:52
[1273] 2015/07/06-14:09:52 =========================================================
[1273] 2015/07/06-14:09:52
[1273] 2015/07/06-14:09:52 [OK]   0 errors, 0 warnings executing configure
```

This example occurs when the user manually launches `ncm-ncd` shortly after the dispatcher decides there is work to do.

Can we add an option to `CAF::Log` so the PID can be recorded, and then modify `ncm-ncd` to use that option when setting up the logger for `ncd.log`?

--- a/src/main/perl/Application.pm
+++ b/src/main/perl/Application.pm
@@ -443,13 +443,14 @@ sub _initialize ($$@) {
     #     $self->set_report_logfile($self->{'LOG'})
     my %logvar = $self->{'CONFIG'}->varlist('^logfile$');
     if (exists $logvar{'logfile'}) {
-        my $logfile=$self->option("logfile");
+        my $logfile = $self->option("logfile");
         if (defined $logfile) {
             # log requested on $logfile. Try to instantiate it and attach
             # it to the reporter.
             my $logflags = 'w';
             $logflags  = 'a' if (defined $self->{'LOG_APPEND'});
             $logflags .= 't' if (defined $self->{'LOG_TSTAMP'});
+            $logflags .= 'p' if (defined $self->{'LOG_PROCID'});
 
             $self->{'LOG'} = CAF::Log->new($logfile, $logflags);
             unless (defined $self->{'LOG'}) {

--- a/src/main/perl/Log.pm
+++ b/src/main/perl/Log.pm
@@ -163,7 +163,7 @@ sub _initialize ($$$)
     $self->{$FILENAME} = $filename;
     $self->{$OPTS} = $options;
 
-    if ($self->{$FILENAME} =~ m{([^/]*).log$}) {
+    if ($self->{$FILENAME} =~ m{([^/]*)\.log$}) {
         $self->{$SYSLOG} = $1;
     }
 

--- a/src/main/perl/Log.pm
+++ b/src/main/perl/Log.pm
@@ -18,6 +18,8 @@ use Readonly;
 Readonly our $FH => 'FH';
 Readonly our $FILENAME => 'FILENAME';
 Readonly my $TSTAMP => 'TSTAMP';
+# variable $PID exists when using 'use English;' (and it is $$)
+Readonly my $PROCID => 'PROCID';
 Readonly my $OPTS => 'OPTS';
 
 our @EXPORT_OK = qw($FILENAME $FH);
@@ -81,6 +83,10 @@ sub close ($)
 
 Prints C<$msg> into the log file.
 
+If C<PROCID> attribute is defined (value is irrelevant),
+the proces id in square brackets (C<[PID]>) and additional
+space are prepended.
+
 If C<TSTAMP> attribute is defined (value is irrelevant),
 a C<YYYY/MM/DD-HH:mm:ss> timestamp and additional space
 are prepended.
@@ -91,18 +97,22 @@ Returns the return value of invocation of FH print method.
 
 =cut
 
-# TODO: use 'if ($self->{$TSTAMP})' rather than only checking if defined
+# TODO: use 'if ($self->{$TSTAMP})' (and PROCID) rather than only checking if defined
 
 sub print ($$)
 {
     my ($self, $msg) = @_;
 
+    if (defined $self->{$PROCID}) {
+        $msg = "[$$] $msg";
+    };
+
     if (defined $self->{$TSTAMP}) {
         # print timestamp the SUE way ;-)
         my ($sec, $min, $hour, $mday, $mon, $year) = localtime(time);
         $msg = sprintf("%04d/%02d/%02d-%02d:%02d:%02d %s",
-                       $year+1900, $mon+1, $mday, $hour, $min, $sec,$msg);
-    }
+                       $year+1900, $mon+1, $mday, $hour, $min, $sec, $msg);
+    };
 
     return $self->{$FH}->print($msg);
 }
@@ -127,6 +137,8 @@ C<$options> is a string with magic letters
 =item w: truncate a loglfile
 
 =item t: generate a timestamp on every print
+
+=item p: add PID
 
 =back
 
@@ -155,19 +167,19 @@ sub _initialize ($$$)
         $self->{$SYSLOG} = $1;
     }
 
-    unless ($self->{$OPTS} =~ /^(w|a)t?$/) {
-
+    unless ($self->{$OPTS} =~ /^[tp]*[wa][tp]*$/) {
         throw_error("Bad options for log ".$self->{$FILENAME}.
                     ": ".$self->{$OPTS});
         return;
     }
 
-    if ($self->{$OPTS} =~ /t/) {
-        $self->{$TSTAMP} = 1;
-    }
+    my %opts = map { $_ => 1 } split( '', $self->{$OPTS});
+
+    $self->{$TSTAMP} = $opts{t};
+    $self->{$PROCID} = $opts{p};
 
     my ($fhmode, $msg);
-    if ($self->{$OPTS} =~ /w/) {
+    if ($opts{w}) {
         # Move old filename away if mode is 'w'.
         rename ($self->{$FILENAME}, $self->{$FILENAME}.'.prev')
             if (-e $self->{$FILENAME});
@@ -200,7 +212,7 @@ Called during garbage collection. Invokes close().
 
 # All Readonly here (and in methods called) might be
 # cleaned up during global cleanup, so use the $_XYZ
-# flavours here (and methods called here). 
+# flavours here (and methods called here).
 sub DESTROY {
     my $self = shift;
     $self->close() if (defined $self->{$_FH});

--- a/src/test/perl/log.t
+++ b/src/test/perl/log.t
@@ -58,14 +58,15 @@ $mock_fh->mock('new', sub {
 $fh_autoflush = 0;
 $fh_new = '';
 my $fn = "target/test/unittest.log";
-$log = CAF::Log->new($fn, 'at');
+$log = CAF::Log->new($fn, 'pat');
 isa_ok($log, 'CAF::Log', 'log is a CAF::Log instance');
 isa_ok($log, 'CAF::Object', 'CAF::Log is a CAF::Object subclass');
 
 is($log->{FILENAME}, $fn, "FILENAME attr set to log filename");
-is($log->{OPTS}, 'at', 'OPTS attr set with initial options');
+is($log->{OPTS}, 'pat', 'OPTS attr set with initial options');
 is($log->{SYSLOG}, 'unittest', 'Filenames ending with .log get SYSLOG attr set to basename');
 is($log->{TSTAMP}, 1, 'presence of t option sets TSTAMP attr to 1');
+is($log->{PROCID}, 1, 'presence of p option sets PROCID attr to 1');
 isa_ok($log->{FH}, 'FileHandle', 'FH attr is teh Filehandle instance');
 is($fh_new, ">> $fn", "Filehandle initialized with append");
 is($fh_autoflush, 1, "autoflush set on Filehandle instance");
@@ -85,6 +86,7 @@ is($log->{FILENAME}, $fn, "FILENAME attr set to log filename");
 is($log->{OPTS}, 'w', 'OPTS attr set with initial options');
 ok(! defined($log->{SYSLOG}), 'Filenames not ending with .log do not get SYSLOG attr set');
 ok(! defined($log->{TSTAMP}), 'absensce of t option does not set TSTAMP attr');
+ok(! defined($log->{PROCID}), 'absensce of p option does not set PROCID attr');
 isa_ok($log->{FH}, 'FileHandle', 'FH attr is teh Filehandle instance');
 is($fh_new, "> $fn", "Filehandle initialized with write");
 is($fh_autoflush, 1, "autoflush set on Filehandle instance");
@@ -96,7 +98,8 @@ ok(-e "$fn.prev", "write mode renames existing logfiles with .prev extension");
 
 =cut
 
-$log->{TSTAMNP} = 0;
+$log->{TSTAMP} = undef;
+$log->{PROCID} = undef;
 $fh_print = '';
 # test print message
 $log->print("mymessage");
@@ -104,10 +107,25 @@ is($fh_print, 'mymessage', "print calls Filehandle->print");
 
 # set TSTAMP, test TSTAMP?
 $log->{TSTAMP} = 1;
+$log->{PROCID} = undef;
 $fh_print = '';
-$log->print("myothermessage");
-like($fh_print, qr{^\d{4}/\d{2}/\d{2}-\d{2}:\d{2}:\d{2} myothermessage$},
+$log->print("myothermessage_t");
+like($fh_print, qr{^\d{4}/\d{2}/\d{2}-\d{2}:\d{2}:\d{2} myothermessage_t$},
      "print with TSTAMP calls Filehandle->print with timestamp prepended");
+
+$log->{TSTAMP} = undef;
+$log->{PROCID} = 1;
+$fh_print = '';
+$log->print("myothermessage_p");
+is($fh_print, "[$$] myothermessage_p",
+     "print with PROCID calls Filehandle->print with pid prepended");
+
+$log->{TSTAMP} = 1;
+$log->{PROCID} = 1;
+$fh_print = '';
+$log->print("myothermessage_tp");
+like($fh_print, qr{^\d{4}/\d{2}/\d{2}-\d{2}:\d{2}:\d{2} \[\d+\] myothermessage_tp$},
+     "print with TSTAMP and PROCID calls Filehandle->print with timestamp and pid prepended");
 
 =pod
 

--- a/src/test/perl/modules/testapp.pm
+++ b/src/test/perl/modules/testapp.pm
@@ -14,13 +14,37 @@ use strict;
 use warnings;
 use CAF::Process;
 use CAF::Application;
-
+use LC::Exception qw (SUCCESS);
 
 our @ISA = qw(CAF::Application);
 
+sub app_options {
+    push(my @array, {
+        NAME => 'logfile=s',
+        HELP => 'log path/filename to use',
+    });
+
+    return \@array;
+}
+
+sub _initialize {
+    my $self = shift;
+
+    $self->{'LOG_TSTAMP'} = 1;
+    $self->{'LOG_PROCID'} = 1;
+
+    # start initialization of CAF::Application
+    if($self->SUPER::_initialize(@_)) {
+        # why is this not the default?
+        $self->set_report_logfile($self->{'LOG'});
+        return SUCCESS;
+    }
+    return;
+}
+
 sub verbose
 {
-    
+
     my ($self, @lines) = @_;
     my $text = join ("", @lines);
     my $fh = $CAF::Reporter::_REP_SETUP->{LOGFILE};

--- a/src/test/perl/test-cafapplication.t
+++ b/src/test/perl/test-cafapplication.t
@@ -1,6 +1,9 @@
+use FindBin qw($Bin);
+use lib "$Bin/modules";
 use strict;
 use warnings;
 use Test::More;
+use testapp;
 
 use Test::MockModule;
 use CAF::Application qw($OPTION_CFGFILE);
@@ -107,5 +110,23 @@ is($app3->{NAME}, 'myname', 'NAME attribute set');
 is($app3->option('myoption'), 'myvalue', 'myoption=myvalue parsed');
 is_deeply($argsref, [qw(not an option)],
           'args array ref does remaining arguments correctly');
+
+# test logging
+# Mock FileHandle new from CAF::Log
+use IO::String;
+my $io = IO::String->new();
+my $mock_fh = Test::MockModule->new('FileHandle');
+$mock_fh->mock('new', $io);
+
+# Pass fake logfile, to enable CAF::Application LOG config
+my $tapp = testapp->new ($0, qw (--quiet --verbose --logfile test/fake/log));
+ok($tapp->{LOG_TSTAMP}, "testapp has TSTAMP enabled");
+ok($tapp->{LOG_PROCID}, "testapp has PROCID enabled");
+$tapp->error("testmessage");
+$io->seek(0);
+like (join('', <$io>), qr{\d+/\d+/\d+-\d+:\d+:\d+ \[\d+\] \[ERROR\] testmessage},
+      "log format as expected logged");
+
+diag explain $tapp;
 
 done_testing();

